flytekit/core/type_engine.py

```diff
@@ -676,12 +676,31 @@ def _make_dataclass_serializable(self, python_val: T, python_type: Type[T]) -> t
         """
         from flytekit.types.directory import FlyteDirectory
         from flytekit.types.file import FlyteFile
+        from flytekit.types.structured import StructuredDataset
 
         # Handle Optional
         if UnionTransformer.is_optional_type(python_type):
-            if python_val is None:
-                return None
-            return self._make_dataclass_serializable(python_val, get_args(python_type)[0])
+
+            def get_expected_type(python_val: T, types: tuple) -> Type[T | None]:
+                if len(set(types) & {FlyteFile, FlyteDirectory, StructuredDataset}) > 1:
+                    raise ValueError("Cannot have two Flyte types in a Union type")
```

Comment:
can we make the error
```
Cannot have more than one Flyte type in the Union when attempting to use the string shortcut. Please specify the full object (e.g. FlyteFile(...)) instead of just passing a string.
```
instead?